daft/datatype.py

```diff
@@ -510,76 +587,406 @@ def to_arrow_dtype(self) -> pa.DataType:
 
-    def _is_list(self) -> builtins.bool:
-        return self._dtype.is_list()
+class IntegerType(NumericType):
+    """Integer subtypes."""
 
-    def _is_boolean(self) -> builtins.bool:
-        return self._dtype.is_boolean()
 
-    def _is_string(self) -> builtins.bool:
-        return self._dtype.is_string()
+class Int8Type(IntegerType):
+    """8-bit integer type."""
 
-    def _is_map(self) -> builtins.bool:
-        return self._dtype.is_map()
+    _dtype = PyDataType.int8()
 
-    def _is_logical_type(self) -> builtins.bool:
-        return self._dtype.is_logical()
 
-    def _is_temporal_type(self) -> builtins.bool:
-        return self._dtype.is_temporal()
+class Int16Type(IntegerType):
+    """16-bit integer type."""
 
-    def _should_cast_to_python(self) -> builtins.bool:
-        # NOTE: This is used to determine if we should cast a column to a Python object type when converting to PyList.
-        # Map is a logical type, but we don't want to cast it to Python because the underlying physical type is a List,
-        # which we can handle without casting to Python.
-        return self._is_logical_type() and not self._is_map()
+    _dtype = PyDataType.int16()
 
-    def __repr__(self) -> str:
-        return self._dtype.__repr__()
 
-    def __eq__(self, other: object) -> builtins.bool:
-        return isinstance(other, DataType) and self._dtype.is_equal(other._dtype)
+class Int32Type(IntegerType):
+    """32-bit integer type."""
 
-    def __reduce__(self) -> tuple:
-        return DataType._from_pydatatype, (self._dtype,)
+    _dtype = PyDataType.int32()
 
-    def __hash__(self) -> int:
-        return self._dtype.__hash__()
+
+class Int64Type(IntegerType):
+    """64-bit integer type."""
+
+    _dtype = PyDataType.int64()
+
+
+class UInt8Type(IntegerType):
+    """Unsigned 8-bit integer type."""
+
+    _dtype = PyDataType.uint8()
+
+
+class UInt16Type(IntegerType):
+    """Unsigned 16-bit integer type."""
+
+    _dtype = PyDataType.uint16()
+
+
+class UInt32Type(IntegerType):
+    """Unsigned 32-bit integer type."""
+
+    _dtype = PyDataType.uint32()
+
+
+class UInt64Type(IntegerType):
+    """Unsigned 64-bit integer type."""
+
+    _dtype = PyDataType.uint64()
+
+
+class FloatType(NumericType):
+    """Floating point subtypes."""
+
+
+class Float32Type(FloatType):
+    """32-bit floating point type."""
+
+    _dtype = PyDataType.float32()
+
+
+class Float64Type(FloatType):
+    """64-bit floating point type."""
+
+    _dtype = PyDataType.float64()
+
+
+class BooleanType(DataType):
+    """Boolean type."""
+
+    _dtype = PyDataType.bool()
+
+
+class StringType(DataType):
+    """UTF-8 encoded string type."""
+
+    _dtype = PyDataType.string()
+
+
+class BinaryType(DataType):
+    """Binary type."""
+
+    _dtype = PyDataType.binary()
+
+
+class FixedSizeBinaryType(DataType):
+    """Fixed-size binary type."""
+
+    @property
+    def size(self) -> int:
+        return self._dtype.get_fixed_size_binary_inner()
```

Comment:
rather than all the inners, you can have a method called `.size()` on the rust dtype. Then that does the switch an returns size for all fixedsize types. So it would have multiple match branches. This way we don't use inner when referring to a property, but we actually use the property name. The downside here is we now have two different switches for the customer, one is switching to find the python class, then switching on the rust side to pull out the property. We could add a .size abstract property to DataType and return AttributeError if the type is incorrect. This would give you the classes while avoid double switches and do the pythonic "ask for forgiveness" thing. It would also remove all the _inner methods and is how you typically handle shared properties with sum types.